libethereum/State.cpp

```diff
@@ -21,720 +21,777 @@
-					contout << " $" << toHex(cache->code());
-				else
-					contout << " $" << (cache ? cache->codeHash() : r[3].toHash<h256>());
-
-				for (auto const& j: mem)
-					if (j.second)
-						contout << std::endl << (delta.count(j.first) ? back.count(j.first) ? " *     " : " +     " : cached.count(j.first) ? " .     " : "       ") << std::hex << nouppercase << std::setw(64) << j.first << ": " << std::setw(0) << j.second ;
-					else
-						contout << std::endl << "XXX    " << std::hex << nouppercase << std::setw(64) << j.first << "";
-			}
-			else
-				contout << " [SIMPLE]";
-			_out << lead << i << ": " << std::dec << (cache ? cache->nonce() : r[0].toInt<u256>()) << " #:" << (cache ? cache->balance() : r[1].toInt<u256>()) << contout.str() << std::endl;
-		}
-	}
-	return _out;
-}
-
-State& dev::eth::createIntermediateState(State& o_s, Block const& _block, unsigned _txIndex, BlockChain const& _bc)
-{
-	o_s = _block.state();
-	u256 const rootHash = _block.stateRootBeforeTx(_txIndex);
-	if (rootHash)
-		o_s.setRoot(rootHash);
-	else
-	{
-		o_s.setRoot(_block.stateRootBeforeTx(0));
-		o_s.executeBlockTransactions(_block, _txIndex, _bc.lastBlockHashes(), *_bc.sealEngine());
-	}
-	return o_s;
+    _out << "--- " << _s.rootHash() << std::endl;
+    std::set<Address> d;
+    std::set<Address> dtr;
+    auto trie = SecureTrieDB<Address, OverlayDB>(
+        const_cast<OverlayDB*>(&_s.m_db), _s.rootHash());
+    for (auto i : trie)
+        d.insert(i.first), dtr.insert(i.first);
+    for (auto i : _s.m_cache)
+        d.insert(i.first);
+
+    for (auto i : d)
+    {
+        auto it = _s.m_cache.find(i);
+        Account* cache = it != _s.m_cache.end() ? &it->second : nullptr;
+        string rlpString = dtr.count(i) ? trie.at(i) : "";
+        RLP r(rlpString);
+        assert(cache || r);
+
+        if (cache && !cache->isAlive())
+            _out << "XXX  " << i << std::endl;
+        else
+        {
+            string lead = (cache ? r ? " *   " : " +   " : "     ");
+            if (cache && r && cache->nonce() == r[0].toInt<u256>() &&
+                cache->balance() == r[1].toInt<u256>())
+                lead = " .   ";
+
+            stringstream contout;
+
+            if ((cache && cache->codeHash() == EmptySHA3) ||
+                (!cache && r && (h256)r[3] != EmptySHA3))
+            {
+                std::map<u256, u256> mem;
+                std::set<u256> back;
+                std::set<u256> delta;
+                std::set<u256> cached;
+                if (r)
+                {
+                    SecureTrieDB<h256, OverlayDB> memdb(
+                        const_cast<OverlayDB*>(&_s.m_db),
+                        r[2].toHash<h256>());  // promise we won't alter the
+                                               // overlay! :)
+                    for (auto const& j : memdb)
+                        mem[j.first] = RLP(j.second).toInt<u256>(),
+                        back.insert(j.first);
```

Comment:
hm, weird, I would expect this line to be indented one level more

Reply:
ah it's two expressions separated by comma in one loop-body-statement, ok, it's the code itself that is weird

Reply:
Wow. This is comma operator. It actually means:
```cpp
for (auto const& j : memdb)
{
    mem[j.first] = RLP(j.second).toInt<u256>();
    back.insert(j.first);
}
```